### PR TITLE
Add Font Ligature Support :GuiRenderLigatures

### PR DESCRIPTION
**Pull request by jgehrig**
**Issue #166:** Font ligatures support request.
**Issue #723:** Font ligatures support request.
**Issue #705:** Underline + Undercurl rendering incorrect.

**Pull #470:** Change printing to add ligature support

First, let me say... Thanks @Nuclearo!
This work would not have happened without your code as reference. The underlying rendering scheme is still being used.

The general approach here is very similar to Pull #470. We update the buffer on a line-by-line basis, and we draw blocks of similarly styled cells together. However, we now use `QTextLayout` and `QGlyphRun` instead of `drawText`. The new API should provide the extra control needed to adjust the spacing of non-monospace fonts and perform ligature detection.

I have tried to structure this code so there is a known-good "old codepath" for `:GuiRenderLigatures 0`. Keeping the code exactly as-is would require large segments to be copy-pasted, so I have tried to decompose rendering tasks into functions. Each function should have code that is almost identical to segments of the old codepath.

**Algorithm:**
1. ShellWidget modifications now `update()` an entire line.
2. On a given line, we look at the first `Cell`'s style.
3. Cell text is appended until a `Cell` style change is detected.
4. Text layout is calculated via `QTextLayout`, resulting in a `QGlyphRun`.
5. The `QGlyphRun` is a list of glyph indices and positions.
6. Ligature are detected from differences in the glyph index of individual characters vs the glyph run.
7. The glyph run values are modified where necessary (cursor on glyph, position, etc)
8. Render the QGlyphRun
9. Back to Step 3 until the line is fully rendered.

**Qt Documentation:**
[QTextLayout](https://doc.qt.io/qt-5/qtextlayout.html)
[QGlyphRun](https://doc.qt.io/qt-5/qglyphrun.html)

### CODE REVIEW

**Comment by jgehrig** [comment]
**Not ready for merge yet.**
I am Posting this early because I assume users will be interested.

There are code cleanup, `update()` performance improvements, and benchmarking still left to do. It would also be nice to have some `ShellWidget` test coverage given such extensive modification.

I may work on integrating this with #690...

File: src/gui/runtime/plugin/nvim_gui_shim.vim [hunk 1]
```diff
@@ -233,6 +233,7 @@ endfunction
 
 command! -nargs=1 GuiScrollBar call s:GuiScrollBar(<args>)
 
+<<<<<<< HEAD
```
Comment (equalsraf) [inline src/gui/runtime/plugin/nvim_gui_shim.vim:236]: Watch out, seems like a merge leftover.

**Comment by jgehrig** [comment]
@equalsraf

I think this one is ready for merge. Let me know if you have any final comments.

The `GuiRenderLigatures 0` state should be very safe, it uses the same logic (refactored) as today.

For `GuiRenderLigatures 1` I am confident in the basic functionality and not aware of any issues. Time for an initial release.

As a side note, the [Clazy Pipeline](https://dev.azure.com/equalsraf/neovim-qt/_build/results?buildId=52&view=logs&jobId=494e7640-935e-5c68-eff6-35d5e173f87c&j=494e7640-935e-5c68-eff6-35d5e173f87c&t=9dd5c925-984f-56cf-8a94-43d4987e307a) already caught a subtle Qt issue :)

**Comment by equalsraf** [comment]
Looking glorious, go forth and press the merge button.

> As a side note, the Clazy Pipeline already caught a subtle Qt issue :)

ooops :D

**Pull request merged**